libs/geometry/src/TangentSpaceMesh.cpp

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright (C) 2023 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <geometry/TangentSpaceMesh.h>
+
+#include <math/vec3.h>
+#include <math/vec4.h>
+#include <utils/Log.h>
+#include <utils/Panic.h>
+
+namespace filament {
+namespace geometry {
+
+using namespace filament::math;
+using Builder = TangentSpaceMesh::Builder;
+using AlgorithmHint = TangentSpaceMesh::AlgorithmHint;
+
+struct TangentSpaceMeshInput {
+    size_t vertexCount = 0;
+    const float3* normals = nullptr;
+    const float4* tangents = nullptr;
+    const float2* uvs = nullptr;
+    const float3* positions = nullptr;
+    const ushort3* triangles16 = nullptr;
+    const uint3* triangles32 = nullptr;
+
+    size_t normalStride = 0;
+    size_t tangentStride = 0;
+    size_t uvStride = 0;
+    size_t positionStride = 0;
+    size_t triangleCount = 0;
+
+    TangentSpaceMesh::AlgorithmHint hint;
+};
+
+struct TangentSpaceMeshOutput {
+    TangentSpaceMesh::AlgorithmHint hint;
+
+    size_t triangleCount = 0;
+    size_t vertexCount = 0;
+
+    float4* tangents = nullptr;
+    float2* uvs = nullptr;
+    float3* positions = nullptr;
+    uint3* triangles32 = nullptr;
+    ushort3* triangles16 = nullptr;
+};
+
+namespace {
+
+const uint8_t NORMALS_BIT = 0x01;
+const uint8_t TANGENTS_BIT = 0x02;
+const uint8_t UVS_BIT = 0x04;
+const uint8_t POSITIONS_BIT = 0x08;
+const uint8_t INDICES_BIT = 0x10;
+
+// Input types
+const uint8_t NORMALS = NORMALS_BIT;
+const uint8_t NORMALS_TANGENTS = NORMALS_BIT | TANGENTS_BIT;
+const uint8_t POSITIONS_INDICES = POSITIONS_BIT | INDICES_BIT;
+const uint8_t NORMALS_UVS_POSITIONS_INDICES = NORMALS_BIT | UVS_BIT | POSITIONS_BIT | INDICES_BIT;
+
+const char* HintToString(AlgorithmHint hint) {
```

Comment:
- return a `std::string_view`
- I think you should name this `getAlgorithmAsString` or even just `to_string` (one of the rare case we break our naming convention for mimicing the stl's).